core/domain/classifier_services.py

```diff
@@ -390,45 +394,160 @@ def get_classifier_training_job_by_id(job_id):
+            raise Exception(
+                'The status change %s to %s is not valid.' % (
+                    initial_status, status))
+
+        classifier_training_job = get_classifier_training_job_by_id(
+            job_id)
+        classifier_training_job.update_status(status)
+        classifier_training_job.validate()
+
+        classifier_training_job_models[index].status = status
+
+    classifier_models.ClassifierTrainingJobModel.put_multi(
+        classifier_training_job_models)
 
-    classifier_training_job_model.status = status
-    classifier_training_job_model.put()
 
 def mark_training_job_complete(job_id):
     """Updates the training job's status to complete.
 
     Args:
         job_id: str. ID of the ClassifierTrainingJob.
     """
-    _update_classifier_training_job_status(job_id,
-                                           feconf.TRAINING_JOB_STATUS_COMPLETE)
+    _update_classifier_training_jobs_status(
+        [job_id], feconf.TRAINING_JOB_STATUS_COMPLETE)
+
+
+def mark_training_jobs_failed(job_ids):
+    """Updates the training job's status to failed.
+
+    Args:
+        job_ids: list(str). list of ID of the ClassifierTrainingJobs.
+    """
+    _update_classifier_training_jobs_status(
+        job_ids, feconf.TRAINING_JOB_STATUS_FAILED)
+
+
+def mark_training_job_pending(job_id):
+    """Updates the training job's status to pending.
+
+    Args:
+        job_id: str. ID of the ClassifierTrainingJob.
+    """
+    _update_classifier_training_jobs_status(
+        [job_id], feconf.TRAINING_JOB_STATUS_PENDING)
+
+
+def _update_job_next_scheduled_check_time(job_id):
+    """Updates the next scheduled check time of job.
+
+    Args:
+        job_id: str. ID of the ClassifierTrainingJob.
+    """
+    classifier_training_job_model = (
+        classifier_models.ClassifierTrainingJobModel.get(job_id))
+
+    if not classifier_training_job_model:
+        raise Exception(
+            'The ClassifierTrainingJobModel corresponding to the job_id '
+            'of the ClassifierTrainingJob does not exist.')
+
+    classifier_training_job_model.next_scheduled_check_time = (
+        datetime.datetime.utcnow() + datetime.timedelta(
+            minutes=feconf.CLASSIFIER_JOB_TTL_MINS))
+    classifier_training_job_model.put()
+
+
+def fetch_next_job():
+    """Gets next job model in the job queue.
+
+    Returns:
+        ClassifierTrainingJob. Domain object of the next training Job.
+    """
+    classifier_training_jobs = []
+    # Initially the cursor for query is set to None.
+    cursor = None
+    valid_jobs = []
+    failed_job_ids = []
+
+    while len(valid_jobs) == 0:
+        classifier_training_jobs, cursor, more = (
+            classifier_models.ClassifierTrainingJobModel.query_training_jobs(
+                cursor))
+        for training_job in classifier_training_jobs:
+            if(training_job.status == (
+                    feconf.TRAINING_JOB_STATUS_PENDING)):
```

Comment:
I think, as a precaution, you should double-check that the TTL has expired too before adding them to the failed jobs list.

Btw, do you know what happens if a job is marked as failed due to timeout, and then later the response comes back from the VM? Is it changed to succeeded in that case? Seems a waste not to...